Update HPX backend

msimberg opened the pull request:
Updates the HPX backend with the following changes:

- Requires HPX 1.7.0 (https://github.com/STEllAR-GROUP/hpx/releases/tag/1.7.0). In 1.7.0 we removed some deprecated functionality that the HPX backend was relying on. This both updates the use of those deprecated features, and uses e.g. new headers and namespaces introduced in 1.7.0.
- Uses local-only headers from HPX.
- Removes the custom executor that was being used in the HPX backend. The default `parallel_executor` is used instead since it now has the same behaviour (and performance) as the custom executor. I've also removed one of the backend implementations as it's not necessary anymore.
- Makes using the same HPX instance thread-safe.
- Adjusts the chunk size internally to be more suitable for HPX. This more or less makes https://github.com/kokkos/kokkos/issues/1866 unnecessary from our perspective. I basically just adjust the chunk size to an integral multiple of the given chunk size, which if I understand correctly is allowed by Kokkos semantics?

Let me know if you have questions about the specific changes, or if you'd prefer separate PRs.

This is still a draft because of some of the fences that this introduces for the asynchronous version of the backend. I'd appreciate your input on what the right places would be for those. I've marked fences that I've introduced because of test failures with `TODO`. I'm also not sure that the behaviour of the HPX backend with named fences is right. For example, in this test: https://github.com/msimberg/kokkos/blob/d1cefbcdbc82c645d54b11857b5924974c61edda/core/unit_test/tools/TestEventCorrectness.hpp#L162-L164, the expected name is "Unnamed instance fence", but a default-constructed HPX instance is the global instance. Since this PR is based on `develop` I expect you most likely have some known failures at the moment. I didn't cross-reference the failures against e.g. an OpenMP build. Do you have a list of currently known failures so that I know what I can safely ignore here at the moment?

DavidPoliakoff commented:
The EventCorrectness test is actually my fault, I didn't test an HPX build. Just change the default instance ID to 1

```c++
  uint32_t m_instance_id = 0; // should be 1
```

And all the other places m_instance_id gets set to `0`, and you'll be in line with other implementations. Apologies. Let me know if you want more info. Basically "instance 0" is used in backends like CUDA to denote the thing that syncs *all* instances (i.e. a cudaDeviceSynchronize). Instance 1 is the default instance.

msimberg commented on core/src/Kokkos_CopyViews.hpp:
```diff
@@ -3084,6 +3092,10 @@ inline void resize(Kokkos::View<T, P...>& v,
+
+  // TODO: Appropriate? Required by the asynchronous HPX backend. Can fail in
+  // TestViewMapping_a.hpp line 939 if there is no fence here.
+  Kokkos::fence();
```
I'll try, thanks. I think the fence being before or after the assignment shouldn't make a difference.

dalg24 replied:
That what I suspect too

msimberg replied:
This seems to work just fine. I still have to get back to the other fences. If you happen to have any insights into them I'd appreciate your thoughts.

One of failures is due to the host deep copy falling back to serial execution without fencing first, which is no good with the HPX backend. The overload of `deep_copy` that takes an instance fences both before and after the actual deep copy, but the one without an instance doesn't fence at all. Would it be reasonable to fence similarly in the non-instance overload?

msimberg commented on core/unit_test/TestDefaultDeviceTypeInit.hpp:
```diff
@@ -86,7 +86,14 @@ char** init_kokkos_args(bool do_threads, bool do_numa, bool do_device,
+    const auto concurrency = std::thread::hardware_concurrency();
+    if (concurrency < nthreads) {
+      nthreads = concurrency;
+    }
```
Fixed, thanks! By the way, I realize these were not all HPX-related changes. Just let me know if they'd be better separated into a different PR.

msimberg commented:
> The EventCorrectness test is actually my fault, I didn't test an HPX build. Just change the default instance ID to 1
> 
> ```c++
>   uint32_t m_instance_id = 0; // should be 1
> ```
> 
> And all the other places m_instance_id gets set to `0`, and you'll be in line with other implementations. Apologies. Let me know if you want more info. Basically "instance 0" is used in backends like CUDA to denote the thing that syncs _all_ instances (i.e. a cudaDeviceSynchronize). Instance 1 is the default instance.

No worries! The default ID is in fact 1 now, I just initialize it in the constructor (I should change that).

The way things are set up currently is that a default-constructed HPX instance does actually block all instances. So that example I linked (https://github.com/msimberg/kokkos/blob/d1cefbcdbc82c645d54b11857b5924974c61edda/core/unit_test/tools/TestEventCorrectness.hpp#L162-L164) is currently a global fence for HPX. It doesn't have to be that way, but that's why I was confused by it expecting an "instance fence" and not a "global fence". Do you have some sort of documentation on the new fencing stuff? If not, I might bother you on slack later in the week to get some clarity on what I should or could do here.

DavidPoliakoff commented:
@msimberg : one of these release cycles I'm just going to do documentation :). No docs yet, feel free to ping on Slack.

And "the default instance fences all instances, but other instances are distinguishable from the default instance" is currently idiosyncratic in Kokkos. Currently either the concept of an instance doesn't really mean anything (a la OpenMP/Serial) or the default instance is actually independently synchronizable (a la CUDA/HIP). If you decide that your semantics will mean that, I'd recommend special-casing the test.

DavidPoliakoff commented on core/unit_test/tools/TestEventCorrectness.hpp:
```diff
@@ -159,6 +159,7 @@ TEST(defaultdevicetype, test_unnamed_instance_fence) {
         {"Unnamed Instance Fence", FencePayload::distinguishable_devices::no,
          root + num_instances}};
     expect_fence_events(expected, [=]() {
+      // TODO: This is the global instance for HPX. Should it not be?
```
A few more thoughts on this. We now have code for partitioning an instance. In CUDA-land, this means creating streams. Consider the following code

```c++
    Kokkos::DefaultExecutionSpace default_space;
    auto instances = Kokkos::Experimental::partition_space(default_space, 1, 1);
    view_type temperature_field1("temperature_one", data_size);
    view_type temperature_field2("temperature_two", data_size);
    auto f1_mirror = Kokkos::create_mirror_view(temperature_field1);
    auto f2_mirror = Kokkos::create_mirror_view(temperature_field2);
    for (int x = 0; x < repeats; ++x) {
      Kokkos::parallel_for(
          "process_temp1", Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(instances[0], 0, data_size),
          KOKKOS_LAMBDA(int i) { temperature_field1(i) -= 1.0f; });
      Kokkos::deep_copy(instances[0], f1_mirror, temperature_field1);
      Kokkos::parallel_for(
          "process_temp2", Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(instances[1], 0, data_size),
          KOKKOS_LAMBDA(int i) { temperature_field2(i) -= 1.0f; });
      Kokkos::deep_copy(instances[1],f2_mirror, temperature_field2);
```

Now, in a CUDA case, if I were to add a fence on "root_space" somewhere in there, it wouldn't fence the instances that were split off (and this is what we considered desirable). In HPX, with your current semantics, fencing the default space fences everything. If I were writing a Kokkos backend, I'd try to make my semantics mirror those of the most-used backends, which in the asynchronous case would be the GPU backends.

It's not a "rule" of Kokkos to my knowledge (@crtrott could say more on that), but if I was playing your hand, I would not have special semantics for the default instance

msimberg replied:
Thanks for the explanation, that makes sense. I haven't really looked at the partitioning so far. What we're using is a flag to the HPX constructor to create an independent instance. It looks like it would be quite easy to implement `partition_space` based on that for HPX. You also don't have to put in much effort to convince me that fencing the default instance should only fence itself. It makes my life easier as well. I think the main reason it fences everything right now is because the default Cuda instance used to do `cudaDeviceSynchronize` (or do I completely misremember...?). Anyways, what it used to do doesn't matter too much. I'll update it to fence itself only.

msimberg replied:
I think the instance ids and fences now behave reasonably. I've left the event correctness test unchanged and it passes so I take that as a sign that it now behaves as you (and I) would hope and expect.

masterleinad commented on core/src/impl/Kokkos_ViewMapping.hpp:
```diff
@@ -2994,7 +2994,14 @@ struct ViewValueFunctor<DeviceType, ValueType, true /* is_scalar */> {
 
-  void destroy_shared_allocation() {}
+  void destroy_shared_allocation() {
+    // TODO: Aren't views supposed to fence before being deleted? Fence all
+    // instances or only one? space is not necessarily the only one using the
+    // view. Does that happen elsewhere? TestScatterView fails with the
+    // asynchronous HPX backend unless there's a fence (e.g.) here. Are all
+    // internal views in scatter views reference counted? Hidden assumptions?
+    space.fence();
```
This is for calling the destructor of the underlying types (that we only assume to be necessary for array types), not for freeing the memory which we assume to be synchronous (in `MemorySpace::deallocate`).

msimberg replied:
I think I was mixing up the behaviour of the Cuda memory space and the host memory space. The Cuda space does synchronize in view destructors, right (at least when the ref count goes to zero)? And the host space does not?

You can ignore the fence here. What happens is that this block triggers a segfault with the asynchronous HPX backend: https://github.com/kokkos/kokkos/blob/e4831441fb213bd29485327b02cf1b54149a58a7/containers/unit_tests/TestScatterView.hpp#L446-L461. `create_scatter_view` spawns an asynchronous `parallel_for` with `ResetDuplicates` which, for big `n` does not complete before the end of the scope. Since there's no fence the parallel for loop segfaults trying to access views that have already been freed. I don't know if it's preferable to have a fence in the test, scatter view constructors, destructor, or when freeing host memory.

msimberg replied:
Any comments on this? I've added a fence over here: https://github.com/kokkos/kokkos/blob/2649edcd57546bf0b5846d35d21b8bbe47e10af9/containers/unit_tests/TestScatterView.hpp#L472 makes the test pass. Would someplace else be more appropriate?

masterleinad replied:
I think we just need a fence in `HostSpace::impl_deallocate()` fencing all the execution spaces that could access host memory.

msimberg replied:
Do you already have functionality to do that generically? `Kokkos::fence()` is too heavy and `Kokkos::DefaultHostExecutionSpace().fence()` is too light. I presume this is something that `ExecSpaceManager` would have to take care of?

masterleinad replied:
No, we don't have that functionality. I think we want something like `Kokkos::fence_host` that only fences the execution spaces for which
```
SpaceAccesibility<ExecutionSpaceType, Kokkos::HostSpace>::accessible
```
is `true`.

masterleinad replied:
We talked about it in the meeting and we would just go with a global fence, i.e. `Kokkos::fence()`.

masterleinad replied:
Would you be willing to just address that in this pull request?

masterleinad replied:
Oh, it looks like you already have it at https://github.com/kokkos/kokkos/pull/4241/files#diff-c4732ec78317e1eb713a2e4eb3397fc76b36b0ea64b84c566ca0f1a29f320e76R311.

msimberg replied:
I pushed the change, but did not test it, hence I didn't comment yet... I suppose you guys are ok with the tradeoffs of the fence at the moment given that you've discussed it? I could imagine it even affecting users who aren't using the HPX backend (e.g. expecting being able to do work on the CPU while things are running asynchronously on the GPU). For us the fence is fine (though a host-only fence would obviously be preferable in the long run).

msimberg replied:
Just to report back, this does fix the view mapping test, but hangs in `impl_shared_alloc`. My guess is that it fences inside a parallel region...

msimberg commented:
I've rebased this again on develop. I think there are two open questions on this that I'd appreciate input on:
1. The host serial optimization for deep copy does not fence before the copy on develop. Is what I've done in this PR OK? See the changes in Kokkos_HostSpace.hpp, Kokkos_HostSpace_deepcopy.{c,h}pp.
2. Host View destructors don't fence any of the host execution spaces on develop. I added a (global) fence before freeing host Views. However, on its own this makes the `impl_shared_alloc` test hang since it attempts to free Views inside a parallel region. Options:
   1. Disable that test for HPX.
   2. Keep adding manual fences before host Views are freed wherever needed.
   3. Use some other mechanism than a parallel for loop for that test.
   4. Don't disable View reference counting for the HPX backend.
   5. ???
I think I would actually prefer 4. if possible, as that would automatically take care of any potential lifetime issues (as long as the views are reference counted in the first place).

msimberg commented:
> Yes, sure, never taking the `std::memcpy` path is fine with me. I just don't know what's better with respect to performance (if we care for small ranges).

Sounds good, I'll do that then. And yes, I'll need to evaluate the performance in more detail at some point. However, for our current use cases I think the performance benefit of avoiding the fence is more important than the performance benefit of doing a serial copy for small ranges. Additionally, there's room to play with the chunk size to make it serial *and* asynchronous for small ranges.

msimberg commented:
I've rebased this on develop now that #4349 is merged.

msimberg commented:
I didn't notice this earlier, but the Pthreads backend hangs consistently and it seems to be due to the new fence in the HostSpace deallocation. I get a backtrace like this at the end of many tests:

```
#0  0x00002aaaab663731 in nanosleep () from /lib64/libpthread.so.0                                                                                                                                                    
#1  0x00000000013dce21 in std::this_thread::sleep_for<long, std::ratio<1l, 1000000l> > (__rtime=...) at /opt/gcc/10.3.0/snos/include/g++/thread:401
#2  0x00000000013dcbf3 in Kokkos::Impl::host_thread_yield (i=43327, mode=Kokkos::Impl::WaitMode::ACTIVE) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_Spinwait.cpp:74
#3  0x000000000040d8ed in Kokkos::Impl::spinwait_while_equal<int> (flag=@0x1b99930: 2, value=2) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_Spinwait.hpp:96
#4  0x00000000013df076 in operator() (__closure=0x7fffffff4ba7) at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.cpp:310
#5  0x00000000013e0601 in Kokkos::Tools::Experimental::Impl::profile_fence_event<Kokkos::Threads, Kokkos::Impl::ThreadsExec::internal_fence(const string&, Kokkos::Impl::fence_is_static)::<lambda()> >(const std::str
ing &, Kokkos::Tools::Experimental::SpecialSynchronizationCases, const struct {...} &) (name=..., reason=Kokkos::Tools::Experimental::GlobalDeviceSynchronization, func=...) at /apps/daint/UES/simbergm/src/kokkos/co
re/src/impl/Kokkos_Profiling.hpp:202
#6  0x00000000013df0c6 in Kokkos::Impl::ThreadsExec::internal_fence (name=..., is_static=Kokkos::Impl::fence_is_static::yes) at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.cpp:322
#7  0x00000000013e09cd in Kokkos::Threads::impl_static_fence (name=...) at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.hpp:647
#8  0x00000000013e04fc in Kokkos::Impl::ThreadsSpaceInitializer::fence (this=0x1c34ca0, name=...) at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.cpp:868
#9  0x00000000013a1fa9 in Kokkos::Impl::ExecSpaceManager::static_fence (this=0x1b57800 <Kokkos::Impl::ExecSpaceManager::get_instance()::space_initializer>, name=...) at /apps/daint/UES/simbergm/src/kokkos/core/src/
impl/Kokkos_Core.cpp:135
#10 0x00000000013a54d5 in Kokkos::Impl::(anonymous namespace)::fence_internal (name=...) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_Core.cpp:568
#11 0x00000000013a8779 in Kokkos::fence (name=...) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_Core.cpp:1098
#12 0x00000000013d4976 in Kokkos::HostSpace::impl_deallocate (this=0x1c1f58c, arg_label=0x1c35e88 "Kokkos::thread_scratch", arg_alloc_ptr=0x1c35e80, arg_alloc_size=2176, arg_logical_size=2048, arg_handle=...) at /a
pps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_HostSpace.cpp:311
#13 0x00000000013d4904 in Kokkos::HostSpace::deallocate (this=0x1c1f58c, arg_label=0x1c35e88 "Kokkos::thread_scratch", arg_alloc_ptr=0x1c35e80, arg_alloc_size=2176, arg_logical_size=2048) at /apps/daint/UES/simberg
m/src/kokkos/core/src/impl/Kokkos_HostSpace.cpp:304
#14 0x00000000013d4b03 in Kokkos::Impl::SharedAllocationRecord<Kokkos::HostSpace, void>::~SharedAllocationRecord (this=0x1c1f550, __in_chrg=<optimized out>) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokk
os_HostSpace.cpp:361
#15 0x00000000013d4b2a in Kokkos::Impl::SharedAllocationRecord<Kokkos::HostSpace, void>::~SharedAllocationRecord (this=0x1c1f550, __in_chrg=<optimized out>) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokk
os_HostSpace.cpp:366
#16 0x00000000013d51ef in Kokkos::Impl::SharedAllocationRecordCommon<Kokkos::HostSpace>::deallocate (arg_rec=0x1c1f550) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_SharedAlloc_timpl.hpp:86
#17 0x00000000013dc770 in Kokkos::Impl::SharedAllocationRecord<void, void>::decrement (arg_record=0x1c1f550) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_SharedAlloc.cpp:290
#18 0x00000000013df564 in Kokkos::Impl::ThreadsExec::execute_resize_scratch (exec=...) at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.cpp:446
#19 0x00000000013df4e1 in Kokkos::Impl::ThreadsExec::execute_serial (func=0x13df51a <Kokkos::Impl::ThreadsExec::execute_resize_scratch(Kokkos::Impl::ThreadsExec&, void const*)>) at /apps/daint/UES/simbergm/src/kokk
os/core/src/Threads/Kokkos_ThreadsExec.cpp:421
#20 0x00000000013df78c in Kokkos::Impl::ThreadsExec::resize_scratch (reduce_size=0, thread_size=0) at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.cpp:496
#21 0x00000000013e01ed in Kokkos::Impl::ThreadsExec::finalize () at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.cpp:746
#22 0x00000000013e0975 in Kokkos::Threads::impl_finalize () at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.hpp:637
#23 0x00000000013e04c8 in Kokkos::Impl::ThreadsSpaceInitializer::finalize (this=0x1c34ca0, all_spaces=false) at /apps/daint/UES/simbergm/src/kokkos/core/src/Threads/Kokkos_ThreadsExec.cpp:862
#24 0x00000000013a1e87 in Kokkos::Impl::ExecSpaceManager::finalize_spaces (this=0x1b57800 <Kokkos::Impl::ExecSpaceManager::get_instance()::space_initializer>, all_spaces=false) at /apps/daint/UES/simbergm/src/kokko
s/core/src/impl/Kokkos_Core.cpp:124
#25 0x00000000013a5406 in Kokkos::Impl::(anonymous namespace)::finalize_internal (all_spaces=false) at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_Core.cpp:560
#26 0x00000000013a86cc in Kokkos::finalize () at /apps/daint/UES/simbergm/src/kokkos/core/src/impl/Kokkos_Core.cpp:1090
#27 0x000000000135cde7 in main (argc=1, argv=0x7fffffff5438) at /apps/daint/UES/simbergm/src/kokkos/core/unit_test/UnitTestMainInit.cpp:55
```

I've seen mentions of the Pthreads backend being replaced with a `std::thread` backend. Is this still a relevant failure? Does someone more knowledgeable about the internals of the Pthreads backend have any suggestions for what to do about this?

Rombur commented:
The Pthreads backend has been replaced. Can you rebase your PR, maybe it will fix the problem.

masterleinad commented:
@msimberg The problem here is tightly related to modifying the `impl_shared_alloc` test. For resizing the scratch space a parallel for is used and that one allocates and deallocates memory. Maybe, there is a way to deactivate the fences for this cases or just implement without a parallel for.

masterleinad commented:
Retest this please.

msimberg commented:
That one red cross looks unrelated to the changes here: https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/7184/pipeline/41.

This is otherwise ready to go from my side. It's been rebased on develop after @masterleinad's changes in #4499 were merged.

masterleinad commented:
Only `sycl.reduce_device_view_team_policy` is failing. We see this test failure occasionally and it is certainly unrelated to any changes in this pull request.

msimberg commented:
Ping? Is there anything blocking this?

msimberg commented on core/src/Kokkos_HPX.hpp:
```diff
@@ -242,51 +245,62 @@ class HPX {
         m_buffer(other.m_buffer),
-        m_future(other.m_future) {}
+        m_future(other.m_future),
+        m_future_mutex(other.m_future_mutex) {}
```
That's a very good question... Thanks for spotting this. The copy constructor has the correct semantics. If fix this.

msimberg replied:
Updated. The copy assignment was also not being tested in `TestHPX_IndependentInstancesInstanceIds.cpp`. I've added a a few tests there.

dalg24 replied:
Tests look good.

I would suggest you do not define neither the copy constructor nor the copy assignment and let the compiler generate them.

msimberg replied:
Good point, that's even better. I've defaulted all four of move/copy constructor/copy assignment operator.

msimberg commented on core/src/Kokkos_HPX.hpp:
```diff
@@ -202,33 +196,42 @@ class thread_buffer {
  public:
-  enum class instance_mode { global, independent };
-  instance_mode m_mode;
+  enum class instance_mode { default_, independent };
```
Because `default` is a keyword. :upside_down_face:. I changed the name because I had the impression that "default" better fits the language that you use for the (...) default instances, rather than global. I'm not set on `default_` though and I might have to come up with a better alternative at some point.

dalg24 replied:
Oh right

dalg24 commented:
Retest this please

msimberg commented:
What about now? Still something blocking this? :) The current failure very much looks unrelated... https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/7320/pipeline/#log-21

msimberg commented:
> Nothing blocking. I was hoping to get one more review but there is no goo reason to keep delaying.

Ah, understandable. Thanks for merging!
